analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContext.cs

```diff
@@ -102,22 +102,34 @@ public void RegisterCodeBlockStartAction<TSyntaxKind>(GeneratedCodeRecognizer ge
 
     public virtual void RegisterCompilationStartAction(Action<SonarCompilationStartAnalysisContext> action) =>
         analysisContext.RegisterCompilationStartAction(
-            c => Execute<SonarCompilationStartAnalysisContext, CompilationStartAnalysisContext>(new(this, c), action, null));
+            c =>
```

Comment:
While this works, I would prefer to move this check to `SonarCompilationStartAnalysisContext.RegisterNodeAction`. Leave this method as is and change SonarCompilationStartAnalysisContext instead like so:
```cs
public sealed class SonarCompilationStartAnalysisContext : SonarAnalysisContextBase<CompilationStartAnalysisContext>
{
    public void RegisterNodeAction<TSyntaxKind>(GeneratedCodeRecognizer generatedCodeRecognizer, Action<SonarSyntaxNodeReportingContext> action, params TSyntaxKind[] syntaxKinds)
        where TSyntaxKind : struct
    {
        if (HasMatchingScope(AnalysisContext.supportedDiagnostics))
        {
            Context.RegisterSyntaxNodeAction(x =>
                Execute<SonarSyntaxNodeReportingContext, SyntaxNodeAnalysisContext>(
                        new(AnalysisContext, x), action, x.Node.SyntaxTree, generatedCodeRecognizer),
                syntaxKinds);
        }
    }

    private static void Execute<TSonarContext, TRoslynContext>(TSonarContext context, Action<TSonarContext> action, SyntaxTree sourceTree, GeneratedCodeRecognizer generatedCodeRecognizer = null)
        where TSonarContext : SonarAnalysisContextBase<TRoslynContext>
    {
        if (context.ShouldAnalyzeTree(sourceTree, generatedCodeRecognizer)
            && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(context.AnalysisContext.supportedDiagnostics, sourceTree)
            && context.AnalysisContext.ShoulAnalyzeRazorFile(sourceTree))
        {
            action(context);
        }
    }
}
```
You have to make some private members internal to do so.

### Why?

In the next steps, we want to capture `ProjectConfiguration()`, `SonarLintXml()`, `IsRazorAnalysisEnabled()`, and so on here. The final method may look like so:

```cs
    public void RegisterNodeAction<TSyntaxKind>(GeneratedCodeRecognizer generatedCodeRecognizer, Action<SonarSyntaxNodeReportingContext> action, params TSyntaxKind[] syntaxKinds)
        where TSyntaxKind : struct
    {
        if (HasMatchingScope(AnalysisContext.supportedDiagnostics))
        {
            var projectConfig = ProjectConfiguration();
            var sonarLintXml = SonarLintXml();
            var isRazorAnalysisEnabled = IsRazorAnalysisEnabled();
            Context.RegisterSyntaxNodeAction(x =>
                Execute<SonarSyntaxNodeReportingContext, SyntaxNodeAnalysisContext>(
                        new(AnalysisContext, x), action, x.Node.SyntaxTree, generatedCodeRecognizer, projectConfig, sonarLintXml, isRazorAnalysisEnabled),
                syntaxKinds);
        }
    }

    private static void Execute<TSonarContext, TRoslynContext>(TSonarContext context, Action<TSonarContext> action, SyntaxTree sourceTree, GeneratedCodeRecognizer generatedCodeRecognizer = null, ProjectConfigReader projectConfig = null, SonarLintXmlReader sonarLintXml = null, bool isRazorAnalysisEnabled = false)
        where TSonarContext : SonarAnalysisContextBase<TRoslynContext>
    {
        // use the captured values here to calculate ShouldAnalyzeTree and ShoulAnalyzeRazorFile
        if (context.ShouldAnalyzeTree(sourceTree, generatedCodeRecognizer)
            && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(context.AnalysisContext.supportedDiagnostics, sourceTree)
            && context.AnalysisContext.ShoulAnalyzeRazorFile(sourceTree))
        {
            action(context);
        }
    }
```

That way, we reduce the number of calls to `ProjectConfiguration()` and friends dramatically and good things will happen.